influxdb/metric.go

```diff
@@ -138,82 +185,101 @@ func (m *Metric) buildQuery(startTime, endTime time.Time) client.Query {
-	if err != nil {
-		return nil, err
-	}
-
 	var filteredPoints []point
 	for _, p := range points {
-		// Once cumulative metrics come in play, we'll filter using lastPoint.
-		// For now we can rely on our constructed query to only fetch the
-		// latest points.
-
-		// For time-aggregated InfluxQL queries with interval i, Influx returns
-		// points with timestamp x, denoting the aggregated value from time
-		// [i, i + x).
-		intervalEnd := p.timestamp.Add(queryInterval)
-
-		// If this time interval hasn't finished accumulating data, wait for
-		// it to complete next time around.
-		if intervalEnd.UnixNano() > endTime.UnixNano() {
+		if m.config.TimeAggregated {
+			// For time-aggregated InfluxQL queries with interval i, Influx
+			// returns points with timestamp x, denoting the aggregated value
+			// from time [i, i + x). Hence, i + x is used to denote its
+			// end time.
+			queryInterval, err := m.config.queryInterval()
+			if err != nil {
+				return nil, err
+			}
+
+			p.timestamp = p.timestamp.Add(queryInterval)
+		}
+
+		// If this (time-aggregated) time interval hasn't finished accumulating
+		// data, wait for it to complete next time around.
+		if p.timestamp.UnixNano() > endTime.UnixNano() {
+			continue
+		}
+
+		// If we have already processed this (cumulative) point, ignore it.
+		if p.timestamp.UnixNano() <= lastPoint.UnixNano() {
 			continue
 		}
 
-		// Update the timestamp on the point as this will be used as the
-		// end timestamp in the corresponding Stackdriver time series.
-		p.timestamp = intervalEnd
 		filteredPoints = append(filteredPoints, p)
 	}
 
 	return filteredPoints, nil
 }
 
-func (m *Metric) convertTimeSeries(points []point) ([]*monitoringpb.TimeSeries, error) {
+func (m *Metric) convertTimeSeries(startTime time.Time, points []point) ([]*monitoringpb.TimeSeries, error) {
 	ts := make([]*monitoringpb.TimeSeries, 0, len(points))
 	for _, p := range points {
-		newP, err := m.convertPoint(p)
+		sdPoint, err := m.convertPoint(startTime, p)
 		if err != nil {
 			return nil, err
 		}
 
 		ts = append(ts, &monitoringpb.TimeSeries{
 			Metric:     &metricpb.Metric{Type: m.StackdriverName()},
 			Resource:   &monitoredres.MonitoredResource{Type: "global"},
-			MetricKind: metricpb.MetricDescriptor_GAUGE,
+			MetricKind: m.metricKind(),
 			ValueType:  metricpb.MetricDescriptor_DOUBLE,
-			Points:     []*monitoringpb.Point{newP},
+			Points:     []*monitoringpb.Point{sdPoint},
 		})
 	}
 	return ts, nil
 }
 
 // convertPoint converts a parsed InfluxDB point into a Stackdriver point.
-func (m *Metric) convertPoint(p point) (*monitoringpb.Point, error) {
+func (m *Metric) convertPoint(startTime time.Time, p point) (*monitoringpb.Point, error) {
 	et, err := ptypes.TimestampProto(p.timestamp)
 	if err != nil {
 		return nil, err
 	}
 
+	interval := &monitoringpb.TimeInterval{
+		EndTime: et,
+	}
+	if m.config.Cumulative {
+		st, err := ptypes.TimestampProto(startTime)
```

Comment:
I think you can just assign to `interval.StartTime` here instead of creating a temporary variable.

Reply:
Done, thanks